Omri93Rajuan/Node-Server-with-JSON-Data
# New version with jwt

JWT Authentication Implementation
Description
This PR implements JWT (JSON Web Token) authentication with secure cookie management. Added login and logout functionality with proper session handling through HTTP-only cookies.
Features Added

JWT token generation and validation
Secure cookie-based authentication
Login endpoint with user validation
Logout endpoint with proper cookie cleanup
TypeScript types for authentication flows

Changes Made
Authentication Service

Implemented JWT token generation using user data
Added secure cookie configuration for token storage
Added token validation middleware

Login Flow
typescriptCopyconst loginUser = async ({ username, password }: SignInDto, res: Response) => {
    try {
        const user = Users.find((item) => 
            item.username === username && 
            comparePassword(password, item.password)
        );

        if (!user) throw new Error("Authentication Error");
        
        const token = generateAuthToken(user);
        res.cookie('auth_token', token, {
            httpOnly: true,
            secure: true,
            sameSite: 'strict',
            maxAge: 24 * 60 * 60 * 1000
        });

        return { message: "Login successful", user: { username: user.username } };
    } catch (error) {
        error.status = 400;
        throw error;
    }
};
Logout Flow
typescriptCopyconst logoutUser = (res: Response) => {
    res.clearCookie('auth_token', {
        httpOnly: true,
        secure: true,
        sameSite: 'strict',
        path: '/'
    });
    return { message: "Logged out successfully" };
};
Routes Added
typescriptCopyrouter.post("/", async (req: LoginRequest, res: Response, next: NextFunction) => {
    try {
        const { username, password } = req.body;
        const user = await loginUser({ username, password }, res);
        res.json(user);
    } catch (error) {
        handleError(res, error);
        next(error);
    }
});

router.post("/logout", async (req: Request, res: Response, next: NextFunction) => {
    try {
        await logoutUser(res);
        res.json({ message: "התנתקת בהצלחה" });
    } catch (error) {
        handleError(res, error);
        next(error);
    }
});
Security Considerations

Cookies are set with httpOnly flag to prevent XSS attacks
Using secure flag to ensure cookies are only sent over HTTPS
Implemented sameSite: 'strict' to prevent CSRF attacks
JWT tokens are stored securely and not exposed to client-side JavaScript
Proper error handling and status codes implemented

Testing Instructions

Test login with valid credentials:

bashCopycurl -X POST http://localhost:3000/auth/login \
  -H "Content-Type: application/json" \
  -d '{"username":"test","password":"test123"}' \
  --cookie-jar cookies.txt

Test protected routes with the cookie:

bashCopycurl http://localhost:3000/protected-route \
  --cookie cookies.txt

=

Closes #XX - Implement JWT Authentication
Closes #YY - Add Logout Functionality

Checklist

 Implemented JWT token generation
 Added secure cookie management
 Implemented login route
 Implemented logout route
 Added error handling
 Added TypeScript types
 Tested all flows
 Updated documentation

Notes

Make sure to set proper CORS configuration if frontend is on a different domain
Environment variables should be set for JWT secret and cookie settings in production
Consider adding refresh token functionality in future updates
